### ISSUE TITLE
'sdf_copy_to()' not properly coping NAs for character and factors

### ISSUE THREAD

**Issue by simonkth** [sparklyr/sparklyr#1503]
When using `spark_apply`, `NA` values in character vectors or factor vectors are lost. When looking closer they appear to be treated as `"NA"`.

Results using `dplyr`:
```r
data <- data.frame(
    grp = rep(c("A", "B", "C"), each = 5),
    logical = c(c(T, T, F, F, F, T, T, F, F), rep(NA, 6)),
    integer = c(1:9, rep(NA, 6)),
    numeric = c(rnorm(9), rep(NA, 6)),
    character = c(letters[1:9], rep(NA, 6)), 
    factor = factor(c(letters[1:9], rep(NA, 6))),
    stringsAsFactors = FALSE
)

library(dplyr)
counts_nas <- function(DF, exclude) {
    DF <- subset(DF, select = colnames(DF)[!colnames(DF) %in% exclude])
    as.data.frame(lapply(DF, function(x) sum(is.na(x))))
}
data %>% group_by(grp) %>% do(counts_nas(., exclude = "grp"))
```
```
# A tibble: 3 x 6
# Groups:   grp [3]
  grp   logical integer numeric character factor
  <chr>   <int>   <int>   <int>     <int>  <int>
1 A           0       0       0         0      0
2 B           1       1       1         1      1
3 C           5       5       5         5      5
```

Results using `sparklyr`:
```r
library(sparklyr)
spark_version <- "2.3.0"
sc <- spark_connect(master = "local", version = spark_version)
data_spark <- sdf_copy_to(sc, data, "data", memory = TRUE)
data_spark %>% spark_apply(
    counts_nas, 
    group_by = "grp", 
    columns = colnames(data_spark)[-1], 
    context = {exclude <- "grp"}
) %>% collect()
```
```
# A tibble: 3 x 6
  grp   logical integer numeric character factor
  <chr>   <int>   <int>   <int>     <int>  <int>
1 B           1       1       1         0      0
2 A           0       0       0         0      0
3 C           5       5       5         0      0
```

Session:
```r
utils::sessionInfo()
```
```
R version 3.5.0 (2018-04-23)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=French_France.1252  LC_CTYPE=French_France.1252    LC_MONETARY=French_France.1252
[4] LC_NUMERIC=C                   LC_TIME=French_France.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] sparklyr_0.8.3 dplyr_0.7.4   

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.16     dbplyr_1.2.1     pillar_1.2.2     compiler_3.5.0   later_0.7.2     
 [6] plyr_1.8.4       bindr_0.1.1      base64enc_0.1-3  tools_3.5.0      digest_0.6.15   
[11] jsonlite_1.5     tibble_1.4.2     nlme_3.1-137     lattice_0.20-35  pkgconfig_2.0.1 
[16] rlang_0.2.0      psych_1.8.4      shiny_1.0.5      DBI_1.0.0        cli_1.0.0       
[21] rstudioapi_0.7   yaml_2.1.19      parallel_3.5.0   bindrcpp_0.2.2   withr_2.1.2     
[26] stringr_1.3.1    httr_1.3.1       rappdirs_0.3.1   rprojroot_1.3-2  grid_3.5.0      
[31] glue_1.2.0       R6_2.2.2         foreign_0.8-70   reshape2_1.4.3   tidyr_0.8.0     
[36] purrr_0.2.4      magrittr_1.5     backports_1.1.2  promises_1.0.1   htmltools_0.3.6 
[41] assertthat_0.2.0 mnormt_1.5-5     mime_0.5         xtable_1.8-2     httpuv_1.4.3    
[46] config_0.3       utf8_1.1.3       stringi_1.2.2    openssl_1.0.1    lazyeval_0.2.1  
[51] broom_0.4.4      crayon_1.3.4    
```

**Comment by harryprince** [comment]
you need to specific `spark_apply` name parameter explicit.

**Comment by simonkth** [comment]
?

**Comment by simonkth** [comment]
@harryprince ?

**Comment by simonkth** [comment]
Please consider:
```r
data_spark %>% spark_apply(
    function(DF, exclude) {
        DF <- subset(DF, select = colnames(DF)[!colnames(DF) %in% exclude])
        as.data.frame(lapply(DF, function(x) sapply(x, function(x) {
            paste(capture.output(print(x)), collapse = "\n")
        })))
    }, 
    group_by = "grp", 
    columns = colnames(data_spark)[-1], 
    context = {exclude <- "grp"}
) %>% collect()
```
```
# A tibble: 15 x 6
   grp   logical   integer numeric         character    factor      
   <chr> <chr>     <chr>   <chr>           <chr>        <chr>       
 1 B     [1] TRUE  [1] 6   [1] -0.08594087 "[1] \"f\""  "[1] \"f\"" 
 2 B     [1] TRUE  [1] 7   [1] -1.016489   "[1] \"g\""  "[1] \"g\"" 
 3 B     [1] FALSE [1] 8   [1] -0.6869393  "[1] \"h\""  "[1] \"h\"" 
 4 B     [1] FALSE [1] 9   [1] 0.6860458   "[1] \"i\""  "[1] \"i\"" 
 5 B     [1] NA    [1] NA  [1] NA          "[1] \"NA\"" "[1] \"NA\""
 6 A     [1] TRUE  [1] 1   [1] -1.209894   "[1] \"a\""  "[1] \"a\"" 
 7 A     [1] TRUE  [1] 2   [1] 0.090604    "[1] \"b\""  "[1] \"b\"" 
 8 A     [1] FALSE [1] 3   [1] 0.6933613   "[1] \"c\""  "[1] \"c\"" 
 9 A     [1] FALSE [1] 4   [1] -0.1582334  "[1] \"d\""  "[1] \"d\"" 
10 A     [1] FALSE [1] 5   [1] -0.857754   "[1] \"e\""  "[1] \"e\"" 
11 C     [1] NA    [1] NA  [1] NA          "[1] \"NA\"" "[1] \"NA\""
12 C     [1] NA    [1] NA  [1] NA          "[1] \"NA\"" "[1] \"NA\""
13 C     [1] NA    [1] NA  [1] NA          "[1] \"NA\"" "[1] \"NA\""
14 C     [1] NA    [1] NA  [1] NA          "[1] \"NA\"" "[1] \"NA\""
15 C     [1] NA    [1] NA  [1] NA          "[1] \"NA\"" "[1] \"NA\""
```

**Comment by yitao-li** [comment]
FYI this has been fixed in sparklyr 1.5
